### PR TITLE
feat: Migration stateless cni

### PR DESCRIPTION
**Pull request by behzad-mir**
This PR contains code changes related to Stateless CNI Migration.
1- There has been code change to use cnireconciler when the managedEndpoint and Migration flag on CNS is set.
2- There has been code changes on CNI GetEndpoint function to fetch the HNS Endpoint corresponded to an IP address to cover the cases for the missing HNSEndpointID in statefile.

- [X] uses [conventional commit messages](https://www.conventionalcommits.org/)
  
- [X] includes documentation
- [ ] adds unit tests
- [X] relevant PR labels added

**Notes**:

### CODE REVIEW

File: cns/cnireconciler/podinfoprovider_test.go [hunk 1]
```diff
@@ -31,7 +31,7 @@ func TestNewCNIPodInfoProvider(t *testing.T) {
 		{
 			name: "good",
 			exec: newCNIStateFakeExec(
-				`{"ContainerInterfaces":{"3f813b02-eth0":{"PodName":"metrics-server-77c8679d7d-6ksdh","PodNamespace":"kube-system","PodEndpointID":"3f813b02-eth0","ContainerID":"3f813b029429b4e41a09ab33b6f6d365d2ed704017524c78d1d0dece33cdaf46","IPAddresses":[{"IP":"10.241.0.17","Mask":"//8AAA=="}]},"6e688597-eth0":{"PodName":"tunnelfront-5d96f9b987-65xbn","PodNamespace":"kube-system","PodEndpointID":"6e688597-eth0","ContainerID":"6e688597eafb97c83c84e402cc72b299bfb8aeb02021e4c99307a037352c0bed","IPAddresses":[{"IP":"10.241.0.13","Mask":"//8AAA=="}]}}}`,
+				`{"ContainerInterfaces":{"3f813b02-eth0":{"PodName":"metrics-server-77c8679d7d-6ksdh","IfName":"eth0","PodNamespace":"kube-system","PodEndpointID":"3f813b02-eth0","ContainerID":"3f813b029429b4e41a09ab33b6f6d365d2ed704017524c78d1d0dece33cdaf46","IPAddresses":[{"IP":"10.241.0.17","Mask":"//8AAA=="}]},"6e688597-eth0":{"PodName":"tunnelfront-5d96f9b987-65xbn","IfName":"eth0","PodNamespace":"kube-system","PodEndpointID":"6e688597-eth0","ContainerID":"6e688597eafb97c83c84e402cc72b299bfb8aeb02021e4c99307a037352c0bed","IPAddresses":[{"IP":"10.241.0.13","Mask":"//8AAA=="}]}}}`,
```
Comment (tamilmani1989) [inline cns/cnireconciler/podinfoprovider_test.go:34]: can we add test for dualstack scenario as well?

File: cns/service/main.go [hunk 2]
```diff
@@ -1227,13 +1227,27 @@ func InitializeCRDState(ctx context.Context, httpRestService cns.HTTPService, cn
...
 				logger.Printf("[Azure CNS] No endpoint state found, skipping initializing CNS state")
+				if cnsconfig.StatelessCNIMigration {
+					logger.Printf("StatelessCNI Migration is enabled")
+					logger.Printf("initializing from Statefull CNI")
+					var endpointState map[string]*restserver.EndpointInfo
+					podInfoByIPProvider, endpointState, err = cnireconciler.NewCNIPodInfoProvider()
+					if err != nil {
+						return errors.Wrap(err, "failed to create CNI PodInfoProvider")
+					}
+					err := endpointStateStore.Write(restserver.EndpointStoreKey, endpointState)
```
Comment (tamilmani1989) [inline cns/service/main.go:1238]: when cni state file will be removed?

File: network/manager.go [hunk 3]
```diff
@@ -504,6 +501,12 @@ func (nm *networkManager) GetEndpointInfo(networkId string, endpointId string) (
 			epInfo.IPAddresses = append(epInfo.IPAddresses, ip.IPv6...)
 
 		}
+		if epInfo.IsEndpointStateIncomplete() {
+			epInfo, err = epInfo.GetEndpointInfoByIP(epInfo.IPAddresses, networkId)
```
Comment (tamilmani1989) [inline network/manager.go:505]: instead of overriding `epInfo` shouldn't we merge with endpoint info obtained from cns otherwise some data will be lost whatever obtained from cns

File: cns/service/main.go (hunk 2 above)
Comment (tamilmani1989) [reply cns/service/main.go:1238]: nvm cni state wont be removed instead detection will be based on cns state file

File: network/endpoint.go [hunk 4]
```diff
@@ -350,3 +353,21 @@ func GetPodNameWithoutSuffix(podName string) string {
...
+	return endpointInfo, nil
+}
+
+// IsEndpointStateComplete returns true if both HNSEndpointID and HostVethName are missing.
+func (epInfo *EndpointInfo) IsEndpointStateIncomplete() bool {
+	if epInfo.HNSEndpointID == "" && epInfo.IfName == "" {
+		return true
+	}
+	return false
+}
```
Comment (jpayne3506) [inline network/endpoint.go:373]: Has the intent of this function changed? Comments are different from code. 

What is the intent for the use in `network/manager.go`?

File: network/endpoint_linux.go [hunk 5]
```diff
@@ -531,3 +531,9 @@ func getDefaultGateway(routes []RouteInfo) net.IP {
 
 	return nil
 }
+
+// GetEndpointInfoByIPImpl returns an endpointInfo that contains corresponding HostVethName.
+// TODO: It needs to be tested to see if HostVethName is required for SingleTenancy
```
Comment (jpayne3506) [inline network/endpoint_linux.go:536]: TODO complete yet?

File: network/endpoint_windows.go [hunk 6]
```diff
@@ -495,3 +496,28 @@ func (ep *endpoint) getInfoImpl(epInfo *EndpointInfo) {
...
+	hnsResponse, err := Hnsv2.GetNetworkByName(networkID)
+	if err != nil {
+		return epInfo, errors.Wrap(err, "HNS Network not found")
+	}
+	hcnEndpoints, err := Hnsv2.ListEndpointsOfNetwork(hnsResponse.Id)
+	if err != nil {
+		return epInfo, errors.Wrap(err, "failed to fetch HNS endpoints for the given network")
+	}
+	for _, hcnEndpoint := range hcnEndpoints {
+		for _, ipConfiguration := range hcnEndpoint.IpConfigurations {
```
Comment (jpayne3506) [inline network/endpoint_windows.go:512]: Is this what handles dualstack?

File: cns/client/client.go [hunk 7]
```diff
@@ -1024,11 +1024,20 @@ func (c *Client) GetHomeAz(ctx context.Context) (*cns.GetHomeAzResponse, error)
 }
 
 // GetEndpoint calls the EndpointHandlerAPI in CNS to retrieve the state of a given EndpointID
-func (c *Client) GetEndpoint(ctx context.Context, endpointID string) (*restserver.GetEndpointResponse, error) {
+func (c *Client) GetEndpoint(ctx context.Context, endpointID, ifName string) (*restserver.GetEndpointResponse, error) {
```
[the remark this answers is not in the document]
Comment (behzad-mir) [reply cns/client/client.go:1027]: We are adding ifName to the call to CNS so it can grab the legacy Endpoint state from CNS which has containerID+Ifname format

File: cns/cnireconciler/podinfoprovider.go [hunk 8]
```diff
@@ -39,15 +44,20 @@ func newCNSPodInfoProvider(endpointStore store.KeyValueStore) (cns.PodInfoByIPPr
...
-	cli := client.New(exec)
+func newCNIPodInfoProvider(exc exec.Interface) (cns.PodInfoByIPProvider, map[string]*restserver.EndpointInfo, error) {
+	cli := client.New(exc)
 	state, err := cli.GetEndpointState()
 	if err != nil {
-		return nil, fmt.Errorf("failed to invoke CNI client.GetEndpointState(): %w", err)
+		return nil, nil, fmt.Errorf("failed to invoke CNI client.GetEndpointState(): %w", err)
+	}
+	for containerID, endpointInfo := range state.ContainerInterfaces {
+		logger.Printf("state dump from CNI: [%+v], [%+v]", containerID, endpointInfo)
```
[the remark this answers is not in the document]
Comment (behzad-mir) [reply cns/cnireconciler/podinfoprovider.go:54]: based on the logs it just put the output within []
I followed the format that we used for other map struct that we had in the code

File: cns/cnireconciler/podinfoprovider.go [hunk 9]
```diff
@@ -39,15 +44,20 @@ func newCNSPodInfoProvider(endpointStore store.KeyValueStore) (cns.PodInfoByIPPr
...
 	}
+	var endpointState map[string]*restserver.EndpointInfo
+	endpointState, err = cniStateToCnsEndpointState(state)
```
[the remark this answers is not in the document]
Comment (behzad-mir) [reply cns/cnireconciler/podinfoprovider.go:57]: We're returning the result from the function in the next line

File: cns/cnireconciler/podinfoprovider.go [hunk 10]
```diff
@@ -68,6 +78,7 @@ func cniStateToPodInfoByIP(state *api.AzureCNIState) (map[string]cns.PodInfo, er
 			podInfoByIP[ipKey] = podInfo
 		}
 	}
+	logger.Printf("podInfoByIP [%+v]", podInfoByIP)
```
[the remark this answers is not in the document]
Comment (behzad-mir) [reply cns/cnireconciler/podinfoprovider.go:81]: will be removed

File: cns/cnireconciler/podinfoprovider.go [hunk 11]
```diff
@@ -101,3 +112,53 @@ func endpointStateToPodInfoByIP(state map[string]*restserver.EndpointInfo) (map[
 	}
 	return podInfoByIP, nil
 }
+
+// cniStateToCnsEndpointState converts an AzureCNIState dumped from a CNI exec
+// into a EndpointInfo map, using the containerID as keys in the map.
+// The map then will be saved on CNS endpoint state
+func cniStateToCnsEndpointState(state *api.AzureCNIState) (map[string]*restserver.EndpointInfo, error) {
```
[the remark this answers is not in the document]
Comment (behzad-mir) [reply cns/cnireconciler/podinfoprovider.go:119]: will add  alog line there

File: cns/cnireconciler/podinfoprovider.go [hunk 12]
```diff
@@ -101,3 +112,53 @@ func endpointStateToPodInfoByIP(state map[string]*restserver.EndpointInfo) (map[
...
+	logger.Printf("Generating CNS ENdpoint State")
+	endpointState := map[string]*restserver.EndpointInfo{}
+	for epID, endpoint := range state.ContainerInterfaces {
+		endpointInfo := &restserver.EndpointInfo{PodName: endpoint.PodName, PodNamespace: endpoint.PodNamespace, IfnameToIPMap: make(map[string]*restserver.IPInfo)}
+		ipInfo := &restserver.IPInfo{}
+		for _, epIP := range endpoint.IPAddresses {
+			if epIP.IP.To4() == nil { // is an ipv6 address
```
[the remark this answers is not in the document]
Comment (behzad-mir) [reply cns/cnireconciler/podinfoprovider.go:126]: yes, because if there is an ip config and it's not V4 then it should be v6?
Twisted if logic though!

File: Makefile [hunk 13]
```diff
@@ -38,7 +38,7 @@ ACN_VERSION				?= $(shell git describe --exclude "azure-ipam*" --exclude "dropgz
 AZURE_IPAM_VERSION		?= $(notdir $(shell git describe --match "azure-ipam*" --tags --always))
 CNI_VERSION				?= $(ACN_VERSION)
 CNI_DROPGZ_VERSION		?= $(notdir $(shell git describe --match "dropgz*" --tags --always))
-CNS_VERSION				?= $(ACN_VERSION)
+CNS_VERSION				?= $(ACN_VERSION)-p
```
Comment (vipul-21) [inline Makefile:41]: Is this intentional ?

File: network/manager.go [hunk 14]
```diff
@@ -412,14 +412,46 @@ func (nm *networkManager) CreateEndpoint(cli apipaClient, networkID string, epIn
...
+	response, err := nm.CnsClient.UpdateEndpoint(context.TODO(), ep.ContainerID, ep.HnsId, ep.HostIfName, ep.IfName)
 	if err != nil {
 		return errors.Wrapf(err, "Update endpoint API returend with error")
 	}
 	logger.Info("Update endpoint API returend ", zap.String("podname: ", response.ReturnCode.String()))
 	return nil
 }
 
+// GetEndpointState will make a call to CNS GetEndpointState API in the stateless CNI mode to fetch the endpointInfo
+func (nm *networkManager) GetEndpointState(networkID, endpointID, ifName string) (*EndpointInfo, error) {
```
Comment (vipul-21) [inline network/manager.go:424]: Can we add UTs for this function ?

File: Makefile (hunk 13 above)
Comment (behzad-mir) [reply Makefile]: yeah added this for my local CNS image to be different than the old one. Removed it from the PR

File: network/manager.go (hunk 14 above)
Comment (behzad-mir) [reply network/manager.go:424]: I have to mock CNS response I believe

Comment (behzad-mir) [reply network/manager.go:424]: I will do this in another PR since it is not related to the Migration code at this point.

File: cns/cnireconciler/podinfoprovider_test.go (hunk 1 above)
Comment (behzad-mir) [reply cns/cnireconciler/podinfoprovider_test.go]: Added.

File: network/manager.go (hunk 3 above)
Comment (behzad-mir) [reply network/manager.go]: This is the first instance that the epinfo struct is getting created. So there is no endpointInfo getting lost from CNI at this point.

File: cns/cnireconciler/podinfoprovider.go [hunk 15]
```diff
@@ -39,15 +44,20 @@ func newCNSPodInfoProvider(endpointStore store.KeyValueStore) (cns.PodInfoByIPPr
...
 }
 
-func newCNIPodInfoProvider(exec exec.Interface) (cns.PodInfoByIPProvider, error) {
-	cli := client.New(exec)
+func newCNIPodInfoProvider(exc exec.Interface) (cns.PodInfoByIPProvider, map[string]*restserver.EndpointInfo, error) {
+	cli := client.New(exc)
 	state, err := cli.GetEndpointState()
 	if err != nil {
-		return nil, fmt.Errorf("failed to invoke CNI client.GetEndpointState(): %w", err)
+		return nil, nil, fmt.Errorf("failed to invoke CNI client.GetEndpointState(): %w", err)
```
[the remark this answers is not in the document]
Comment (behzad-mir) [reply cns/cnireconciler/podinfoprovider.go:51]: I just followed teh existing code on CNS. Didn't want to add a new error format.

File: network/endpoint_windows.go (hunk 6 above)
Comment (behzad-mir) [reply network/endpoint_windows.go]: nope this will get hns ID from HNS if we don;t have the state

File: cns/client/client.go (hunk 7 above)
[the remark this answers is not in the document]
Comment (rbtr) [reply cns/client/client.go:1027]: is this a breaking change?

File: cns/restserver/ipam.go [hunk 16]
```diff
@@ -1068,6 +1084,11 @@ func (service *HTTPRestService) GetEndpointHelper(endpointID string) (*EndpointI
 		logger.Warnf("[GetEndpointState] Found existing endpoint state for container %s", endpointID)
 		return endpointInfo, nil
 	}
+	legacyEndpointID := endpointID[:ContainerIDLength] + "-" + req.IFName
```
Comment (tamilmani1989) [inline cns/restserver/ipam.go:1087]: can you add a comment on why this change is required?

File: network/manager.go (hunk 14 above)
Comment (tamilmani1989) [reply network/manager.go:424]: can you file workitem and mention as TODO in code

File: cni/linux.Dockerfile [hunk 17]
```diff
@@ -13,6 +13,9 @@ RUN GOOS=$OS CGO_ENABLED=0 go build -a -o /go/bin/azure-vnet-telemetry -trimpath
 RUN GOOS=$OS CGO_ENABLED=0 go build -a -o /go/bin/azure-vnet-ipam -trimpath -ldflags "-X main.version="$VERSION"" -gcflags="-dwarflocationlists=true" cni/ipam/plugin/main.go
 RUN GOOS=$OS CGO_ENABLED=0 go build -a -o /go/bin/azurecni-stateless -trimpath -ldflags "-X main.version="$VERSION"" -gcflags="-dwarflocationlists=true" cni/network/stateless/main.go
 
+FROM scratch as bins
```
[the remark this answers is not in the document]
Comment (behzad-mir) [reply cni/linux.Dockerfile:16]: Stateless CNI binary was not getting built without this line. It was suggested by @rbtr

File: cns/cnireconciler/podinfoprovider.go (hunk 9 above)
[the remark this answers is not in the document]
Comment (behzad-mir) [reply cns/cnireconciler/podinfoprovider.go:57]: need to add a flag to the main reconciler function.

File: cns/cnireconciler/podinfoprovider.go [hunk 18]
```diff
@@ -39,15 +44,20 @@ func newCNSPodInfoProvider(endpointStore store.KeyValueStore) (cns.PodInfoByIPPr
 	}), nil
 }
 
-func newCNIPodInfoProvider(exec exec.Interface) (cns.PodInfoByIPProvider, error) {
-	cli := client.New(exec)
+func newCNIPodInfoProvider(exc exec.Interface) (cns.PodInfoByIPProvider, map[string]*restserver.EndpointInfo, error) {
```
[the remark this answers is not in the document]
Comment (behzad-mir) [reply cns/cnireconciler/podinfoprovider.go:47]: added

File: network/endpoint_linux.go (hunk 5 above)
Comment (behzad-mir) [reply network/endpoint_linux.go:536]: WorkItem is added

File: cns/restserver/ipam.go [hunk 19]
```diff
@@ -1068,6 +1074,11 @@ func (service *HTTPRestService) GetEndpointHelper(endpointID string) (*EndpointI
 		logger.Warnf("[GetEndpointState] Found existing endpoint state for container %s", endpointID)
 		return endpointInfo, nil
 	}
+	legacyEndpointID := endpointID[:ContainerIDLength] + "-" + InterfaceName
```
Comment (tamilmani1989) [inline cns/restserver/ipam.go:1077]: can you add comment on this change?

File: cns/restserver/ipam.go [hunk 20]
```diff
@@ -27,6 +27,12 @@ var (
 	ErrNoNCs                  = errors.New("no NCs found in the CNS internal state")
 	ErrOptManageEndpointState = errors.New("CNS is not set to manage the endpoint state")
 	ErrEndpointStateNotFound  = errors.New("endpoint state could not be found in the statefile")
+	ErrExistingIpconfigFound  = errors.New("Found existing ipconfig for infra container")
```
Comment (tamilmani1989) [inline cns/restserver/ipam.go:30]: can we remove this error if not used?

**Review by tamilmani1989 — APPROVED** [review]

File: cns/restserver/ipam.go (hunk 19 above)
Comment (behzad-mir) [reply cns/restserver/ipam.go:1079]: added

File: cns/cnireconciler/podinfoprovider.go (hunk 15 above)
[the remark this answers is not in the document]
Comment (behzad-mir) [reply cns/cnireconciler/podinfoprovider.go:50]: Will keep it with current format for now.

File: network/endpoint.go (hunk 4 above)
Comment (behzad-mir) [reply network/endpoint.go]: nice catch

File: cns/restserver/ipam.go [hunk 21]
```diff
@@ -1008,9 +1011,22 @@ func (service *HTTPRestService) EndpointHandlerAPI(w http.ResponseWriter, r *htt
...
 func (service *HTTPRestService) GetEndpointHandler(w http.ResponseWriter, r *http.Request) {
 	logger.Printf("[GetEndpointState] GetEndpoint for %s", r.URL.Path)
-
+	var req cns.EndpointRequest
+	err := service.Listener.Decode(w, r, &req)
 	endpointID := strings.TrimPrefix(r.URL.Path, cns.EndpointPath)
-	endpointInfo, err := service.GetEndpointHelper(endpointID)
+	logger.Request(service.Name, &req, err)
+	// Check if the request is valid
+	if err != nil || req.IFName == "" {
```
[the remark this answers is not in the document]
Comment (behzad-mir) [reply cns/restserver/ipam.go]: Old CNI will not call to this function and Ifname has been removed

File: cns/restserver/ipam.go (hunk 16 above)
Comment (behzad-mir) [reply cns/restserver/ipam.go]: added

File: cns/restserver/ipam.go (hunk 20 above)
Comment (behzad-mir) [reply cns/restserver/ipam.go]: removed

File: cns/service/main.go [hunk 22]
```diff
@@ -1234,17 +1234,14 @@ func InitializeCRDState(ctx context.Context, httpRestService cns.HTTPService, cn
 	switch {
 	case cnsconfig.ManageEndpointState:
 		logger.Printf("Initializing from self managed endpoint store")
-		podInfoByIPProvider, err = cnireconciler.NewCNSPodInfoProvider(httpRestServiceImplementation.EndpointStateStore) // get reference to endpoint state store from rest server
+		podInfoByIPProvider, err = InitializeStateFromCNS(cnsconfig, httpRestServiceImplementation.EndpointStateStore)
```
Comment (rbtr) [inline cns/service/main.go:1237]: instead of it being the responsibility of the restore state block to _also_ do the state migration, we should conditionally do the migration first. 
all of this stays unchanged, but in a block above, we can:
```go
if cnsconfig.EnableStateMigration {
    // if CNS statefile exists then:
        // short circuit, no need to do anything
    // else then:
        // construct CNI PodInfoProvider
        // convert PodInfoByIP to state format
        // write CNS statefile
}
// proceed to reconcile from CNS state as normal
```

Comment (rbtr) [reply cns/service/main.go:1237]: this is a major blocker for me

**Review by rbtr — CHANGES_REQUESTED** [review]

File: cns/restserver/ipam.go [hunk 23]
```diff
@@ -1068,6 +1073,14 @@ func (service *HTTPRestService) GetEndpointHelper(endpointID string) (*EndpointI
 		logger.Warnf("[GetEndpointState] Found existing endpoint state for container %s", endpointID)
 		return endpointInfo, nil
 	}
+	// This part is a temprory fix if we have endpoint states belong to CNI version 1.4.X on Windows since the states don't have the containerID
```
[the remark this answers is not in the document]
Comment (behzad-mir) [reply cns/restserver/ipam.go:1076]: SO the issue is within CNI 1.4.X and the only info it gioves us is the [first-8-containerID-digits]-eth0 and also the IP address. So the data is not there at CNS to convert them. We need this special case for 1.4.x!

File: network/manager.go [hunk 24]
```diff
@@ -420,6 +420,39 @@ func (nm *networkManager) UpdateEndpointState(ep *endpoint) error {
 	return nil
 }
 
+// GetEndpointState will make a call to CNS GetEndpointState API in the stateless CNI mode to fetch the endpointInfo
+// TODO unit tests need to be added, WorkItem: 26606939
```
[the remark this answers is not in the document]
Comment (behzad-mir) [reply network/manager.go:424]: So this is not related to Stateless CNI migration PR and is related to Stateless CNI operation and will be covered in a future PR. @tamilmani1989 asked me to add work item in here.

File: network/endpoint_linux.go [hunk 25]
```diff
@@ -531,3 +531,9 @@ func getDefaultGateway(routes []RouteInfo) net.IP {
 
 	return nil
 }
+
+// GetEndpointInfoByIPImpl returns an endpointInfo that contains corresponding HostVethName.
+// TODO: It needs to be tested to see if HostVethName is required for SingleTenancy, WorkItem: 26606939
```
[the remark this answers is not in the document]
Comment (behzad-mir) [reply network/endpoint_linux.go:536]: same as above

File: cns/restserver/ipam.go (hunk 23 above)
[the remark this answers is not in the document]
Comment (rbtr) [reply cns/restserver/ipam.go:1076]: the way I previously handled this was via the `Key()` abstraction on the `PodInfo`, so that the business logic didn't actually need to know that there were different identifier schemas in the backend. Would something like that work here?

File: network/endpoint.go [hunk 26]
```diff
@@ -350,3 +353,21 @@ func GetPodNameWithoutSuffix(podName string) string {
...
+
+// GetEndpointInfoByIP returns an EndpointInfo with complete state when HNS Endpoint ID or HostVeth Name are missing on the CNS state.
+func (epInfo *EndpointInfo) GetEndpointInfoByIP(ipAddresses []net.IPNet, networkID string) (*EndpointInfo, error) {
+	// Call the platform implementation.
+	endpointInfo, err := epInfo.GetEndpointInfoByIPImpl(ipAddresses, networkID)
+	if err != nil {
+		return nil, err
+	}
+	return endpointInfo, nil
+}
```
[the remark this answers is not in the document]
Comment (behzad-mir) [reply network/endpoint.go:365]: correct.

File: cns/restserver/ipam.go (hunk 23 above)
[the remark this answers is not in the document]
Comment (behzad-mir) [reply cns/restserver/ipam.go:1076]: We can add a Key function but then this special case should be implemented there as well. Since stateless CNI will always query via full containerID and we have to build this special case from it.

File: cns/service/main.go (hunk 22 above)
Comment (behzad-mir) [reply cns/service/main.go:1237]: Addressed in the latest commit.

File: cns/configuration/configuration.go [hunk 27]
```diff
@@ -56,6 +56,7 @@ type CNSConfig struct {
 	UseHTTPS                    bool
 	WatchPods                   bool `json:"-"`
 	WireserverIP                string
+	StateMigration              bool
```
[the remark this answers is not in the document]
Comment (behzad-mir) [reply cns/configuration/configuration.go]: Addressed in the latest commit.

File: network/endpoint.go [hunk 28]
```diff
@@ -350,3 +353,11 @@ func GetPodNameWithoutSuffix(podName string) string {
 	logger.Info("Pod name after splitting based on", zap.Any("nameSplit", nameSplit))
 	return strings.Join(nameSplit, "-")
 }
+
+// IsEndpointStateInComplete returns true if both HNSEndpointID and HostVethName are missing.
+func (epInfo *EndpointInfo) IsEndpointStateIncomplete() bool {
+	if epInfo.HNSEndpointID == "" && epInfo.IfName == "" {
+		return true
+	}
+	return false
```
Comment (rbtr) [inline network/endpoint.go:362]:
```suggestion
	return epInfo.HNSEndpointID == "" && epInfo.IfName == ""
```

File: cns/service/main.go [hunk 29]
```diff
@@ -1516,3 +1522,17 @@ func createOrUpdateNodeInfoCRD(ctx context.Context, restConfig *rest.Config, nod
 
 	return nil
 }
+
+// PopulateCNSEndpointState initilizes CNS Endpoint State by Migrating the CNI state.
+func PopulateCNSEndpointState(endpointStateStore store.KeyValueStore) error {
```
Comment (rbtr) [inline cns/service/main.go:1527]: can be private?
```suggestion
func populateCNSEndpointState(endpointStateStore store.KeyValueStore) error {
```

File: cns/cnireconciler/podinfoprovider.go [hunk 30]
```diff
@@ -101,3 +104,57 @@ func endpointStateToPodInfoByIP(state map[string]*restserver.EndpointInfo) (map[
 	}
 	return podInfoByIP, nil
 }
+
+// MigrateCNISate returns an endpoint state of CNS by reading the CNI state file
+func MigrateCNISate() (map[string]*restserver.EndpointInfo, error) {
+	return migrateCNISate(exec.New())
+}
+
+func migrateCNISate(exc exec.Interface) (map[string]*restserver.EndpointInfo, error) {
```
Comment (rbtr) [inline cns/cnireconciler/podinfoprovider.go:113]: these are fine in this package but should probably be in a `statemigrator.go` or something similar, to make it clear that they are unrelated to the PodInfoProvider

**Review by rbtr — CHANGES_REQUESTED** [review]

**Review by rbtr — APPROVED** [review]
remaining suggestions are not blocking
